### PR TITLE
[WIP] Execute custom code on Startup

### PR DESCRIPTION
**Pull request by lgeiger**
A (not very fancy) solution for https://github.com/nteract/hydrogen/issues/112.

Execute custom code when the kernel is started.

### CODE REVIEW

**Comment by n-riesco** [comment]
@rgbkrk I'm reviewing this PR. Please, don't merge it for the time being.

**Comment by n-riesco** [comment]
@lgeiger Here's my review:

* `main.coffee` has grown too large, we should aim at moving functionality out of `main.coffee`; by this I mean:

  * If there's a choice between adding functionality to `main.coffee` or to any other file, it's preferable to add the functionality to other files.

  * For example, functionality recently added, like the Inspector, should be moved into its own file (similarly to the autocomplete provider or watch view).

* In this case, the PR touches `main.coffee`, and `kernel-manager.coffee`. There is no need to touch `main.coffee`, it is possible to limit all the changes to `kernel-manager.coffee`. I think, though, that since the startup code is only run on kernel construction, this PR should be limited to changing `kernel.coffee`.

* There is also the question whether the startup code should be specific to a language or to a kernel. For example, for Javascript, there is one kernel that won't support magics, and the other two don't support the same magics.

File: lib/kernel-manager.coffee [hunk 1]
```diff
@@ -176,3 +177,6 @@ module.exports = KernelManager =
 
     destroy: ->
         _.forEach @runningKernels, (kernel) -> kernel.destroy()
+
+    getStartupCode: (language) ->
+        return @getConfigJson('startupCode')[language] + ' \n'
```
Comment (n-riesco) [inline lib/kernel-manager.coffee]: If no startup script is defined, this function returns the string 'undefined\n'.

Comment (lgeiger) [inline lib/kernel-manager.coffee]: Thanks for noticing! I added the line break at some point and didn't thought about this.

**Comment by lgeiger** [comment]
1. addressed in #194 

2. I try to figure out a way to do this

3. I didn't thought about this. For me it would make sense to have a kernel specific startup code.

**Comment by n-riesco** [comment]
On 15/03/16 11:20, Lukas Geiger wrote:
>  2. I try to figure out a way to do this

I think moving all the relevant code to the kernel constructor (or to a method called by the constructor) should be enough. Please, let me know if there are any difficulties.

**Comment by lgeiger** [comment]
The easiest thing for me to do is moving it to `startKernel` in `kernel-manager.coffee`. It would reduce the code to just four lines. I'll just have to do some testing.

**Comment by lgeiger** [comment]
Can we use use the `kernelInfo.display_name` instead of `kernelInfo.language` to achieve point 3?

**Comment by n-riesco** [comment]
On 15/03/16 11:52, Lukas Geiger wrote:
> The easiest thing for me to do is moving it to |startKernel| in |kernel-manager.coffee|. It would reduce the code to just four lines. I'll just have to do some testing.

SGTM

**Comment by n-riesco** [comment]
On 15/03/16 12:10, Lukas Geiger wrote:
> Can we use use the |kernelInfo.display_name| instead of |kernelInfo.language| to achieve point 3?

I think that's the easiest option I can see at the moment.

**Comment by lgeiger** [comment]
@n-riesco can you test it? I don't have multiple kernels for the same language installed.

**Comment by n-riesco** [comment]
@lgeiger It's working for me. I've tested it with `startupCode` configured as:

```json
{"IPython 2": "test = \"IPython 2\"", "IPython 3": "test = \"IPython 3\""}
```

and `kernelspec`:

```json
{
    "kernelspecs": {
        "ipython2": {
            "spec": {
                "display_name": "IPython 2",
                "env": {},
                "argv": [
                    "ipython",
                    "kernel",
                    "--no-secure",
                    "--colors=NoColor",
                    "-f",
                    "{connection_file}"
                ],
                "language": "python"
            }
        },
        "ipython3": {
            "spec": {
                "display_name": "IPython 3",
                "env": {},
                "argv": [
                    "ipython3",
                    "kernel",
                    "--no-secure",
                    "--colors=NoColor",
                    "-f",
                    "{connection_file}"
                ],
                "language": "python"
            }
        }
    }
}
```

**Comment by n-riesco** [comment]
Shall we merge?

**Comment by rgbkrk** [comment]
> I'm reviewing this PR. Please, don't merge it for the time being.

Thanks for the heads up as well as the thorough review.

**Pull request merged**